modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java

```diff
@@ -50,28 +61,40 @@
      * @param changer Configuration changer.
      * @param listenOnly Only adding listeners mode, without the ability to get or update the property value.
      * @param readOnly Value cannot be changed.
-     * @param injectedNameField Configuration field with {@link InjectedName}.
      */
     public DynamicProperty(
             List<String> prefix,
             String key,
             RootKey<?, ?> rootKey,
             DynamicConfigurationChanger changer,
             boolean listenOnly,
-            boolean readOnly,
-            boolean injectedNameField
+            boolean readOnly
     ) {
-        super(prefix, key, rootKey, changer, listenOnly);
+        super(
+                INJECTED_NAME.equals(key) || INTERNAL_ID.equals(key) ? prefix : appendKey(prefix, key),
+                key,
+                rootKey,
+                changer,
+                listenOnly
+        );
 
         this.readOnly = readOnly;
-        this.injectedNameField = injectedNameField;
+
+        this.injectedNameField = INJECTED_NAME.equals(key);
+        this.internalIdField = INTERNAL_ID.equals(key);
     }
 
     /** {@inheritDoc} */
     @Override
     public T value() {
         if (injectedNameField) {
+            // In this case "refreshValue()" is not of type "T", but an "InnerNode" holding it instead.
```

Comment:
I see that it's guarded, but external invariants can be broken (especially in large codebase such as our configuration), probably an assertion would make sense?

Reply:
If it is not used correctly, an **ClassCastException** will be thrown, if it makes sense to add **assert**?

Reply:
We discussed it in slack - we'll leave it as it is.